sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/write/DocumentWriteService.java

```diff
@@ -409,6 +416,94 @@ private Uni<ResultSet> deleteDocumentInternal(
                     requestInfo.getStargateBridge(), query, context.nested("ASYNC DELETE")));
   }
 
+  /**
+   * Deletes a given dead leaves for a document.
+   *
+   * @param keyspace Keyspace to delete a dead leaves from.
+   * @param collection Collection the document belongs to.
+   * @param documentId Document ID.
+   * @param microsTimestamp Micros timestamp to use in delete queries.
+   * @param deadLeaves A map of JSON paths (f.e. $.some.path) to dead leaves to delete.
+   * @param context Execution content for profiling.
+   * @return Single containing the {@link ResultSet} of the batch execution.
+   */
+  public Uni<ResultSet> deleteDeadLeaves(
+      String keyspace,
+      String collection,
+      String documentId,
+      long microsTimestamp,
+      Map<String, Set<DeadLeaf>> deadLeaves,
+      ExecutionContext context) {
+    return Uni.createFrom()
+        .item(
+            () -> {
+              List<QueryOuterClass.BatchQuery> queries = new ArrayList<>();
+              for (Map.Entry<String, Set<DeadLeaf>> entry : deadLeaves.entrySet()) {
+                String path = entry.getKey();
+                Set<DeadLeaf> leaves = entry.getValue();
+
+                getDeadLeavesQueryBuilders(path, leaves)
+                    .forEach(
+                        builder -> {
+                          queries.add(
+                              builder.buildAndBind(
+                                  keyspace, collection, documentId, microsTimestamp));
+                        });
+              }
+              return queries;
+            })
+        .flatMap(
+            boundQueries ->
+                executeBatch(
+                    requestInfo.getStargateBridge(),
+                    boundQueries,
+                    context.nested("ASYNC DOCUMENT CORRECTION")));
+  }
+
+  // creates needed query builders for one path of dead leaves
+  private List<AbstractDeleteQueryBuilder> getDeadLeavesQueryBuilders(
+      String path, Set<DeadLeaf> leaves) {
+    List<String> pathParts = PATH_SPLITTER.splitToList(path);
+    if (pathParts.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    List<String> pathToDelete = pathParts.subList(1, pathParts.size());
+    boolean deleteArray = false;
+    boolean deleteAll = false;
+    List<String> keysToDelete = new ArrayList<>(leaves.size());
+    for (DeadLeaf deadLeaf : leaves) {
+      if (DeadLeaf.STAR_LEAF.equals(deadLeaf)) {
+        deleteAll = true;
+      } else if (DeadLeaf.ARRAY_LEAF.equals(deadLeaf)) {
+        deleteArray = true;
+      } else {
+        keysToDelete.add(deadLeaf.getName());
+      }
+    }
+
+    List<AbstractDeleteQueryBuilder> builders = new ArrayList<>();
+
+    // TODO Eric if we have delete all, why would we add this one below
```

Comment:
This is a great point...

You could probably just rewrite this to
if (deleteAll) {
    // handle delete all
} else {
    if (deleteArray) {
    }
    if (keysToDelete) {
    }
}

Since if deleteAll is true you don't have to do the other two (but those other two could both fire off in theory)